FK_solver/src/FK_solver.hpp

```diff
@@ -0,0 +1,282 @@
+// #ifndef HEAT_NON_LINEAR_HPP
+// #define HEAT_NON_LINEAR_HPP
+#ifndef FK_SOLVER_HPP
+#define FK_SOLVER_HPP
+
+#include <deal.II/base/conditional_ostream.h>
+#include <deal.II/base/quadrature_lib.h>
+
+#include <deal.II/distributed/fully_distributed_tria.h>
+
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_tools.h>
+
+#include <deal.II/fe/fe_simplex_p.h>
+#include <deal.II/fe/fe_system.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/fe/fe_values_extractors.h>
+#include <deal.II/fe/mapping_fe.h>
+
+#include <deal.II/grid/grid_in.h>
+
+#include <deal.II/lac/solver_cg.h>
+#include <deal.II/lac/trilinos_precondition.h>
+#include <deal.II/lac/trilinos_sparse_matrix.h>
+#include <deal.II/lac/vector.h>
+
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/numerics/matrix_tools.h>
+#include <deal.II/numerics/vector_tools.h>
+
+#include <fstream>
+#include <iostream>
+
+using namespace dealii;
+
+// Class representing the non-linear diffusion problem.
+class FK_solver
+{
+public:
+  // Physical dimension (1D, 2D, 3D)
+  static constexpr unsigned int dim = 3;
+
+
+  // Function for the alpha coefficient.
+  class FunctionAlpha : public Function<dim>
+  {
+  public:
+    virtual double
+    value(const Point<dim> & /*p*/,
+          const unsigned int /*component*/ = 0) const override
+    {
+      return 0.9;
+    }
+  };
+
+  // Function for the forcing term.
+  class ForcingTerm : public Function<dim>
+  {
+  public:
+    virtual double
+    value(const Point<dim> & /*p*/,
+          const unsigned int /*component*/ = 0) const override
+    {
+      return 0.0;
+    }
+  };
+
+  // Function of the matrix D
+  class FunctionD : public Function<dim>
+  {
+  public:
+    virtual void
+    matrix_value(const Point<dim> & /*p*/,
+                 FullMatrix<double> &values) const // override
```

Comment:
Mathematically speaking, $D$ is a **tensor**, which is represented by a matrix for a given coordinate system. This is the same (admittedly subtle) difference that exists between an n-dimensional vector (in the physical sense) and its representation through the n-by-1 matrix of its components with respect to a given coordinate system.

`deal.II` uses the class template `Tensor` to represent such objects: `Tensor<1, dim>` for `dim`-dimensional physical vectors (e.g. gradients, or velocities) and `Tensor<2, dim>` for `dim`-dimensional tensors of rank 2.

I suggest you use `Tensor<2,dim>` in place of `FullMatrix<double>`. On top of many semantical benefits, this will allow you to compute $D\nabla\varphi_j$ with the multiplication operator `*`, without the need to loop over components.